baseapp/abci.go

```diff
@@ -541,194 +790,165 @@ func (app *BaseApp) halt() {
-	if app.snapshotManager == nil {
-		return abci.ResponseLoadSnapshotChunk{}
-	}
-	chunk, err := app.snapshotManager.LoadChunk(req.Height, req.Format, req.Chunk)
-	if err != nil {
-		app.logger.Error(
-			"failed to load snapshot chunk",
-			"height", req.Height,
-			"format", req.Format,
-			"chunk", req.Chunk,
-			"err", err,
-		)
-		return abci.ResponseLoadSnapshotChunk{}
+func handleQueryP2P(app *BaseApp, path []string) *abci.ResponseQuery {
+	// "/p2p" prefix for p2p queries
+	if len(path) < 4 {
+		return sdkerrors.QueryResult(errorsmod.Wrap(sdkerrors.ErrUnknownRequest, "path should be p2p filter <addr|id> <parameter>"), app.trace)
 	}
-	return abci.ResponseLoadSnapshotChunk{Chunk: chunk}
-}
 
-// OfferSnapshot implements the ABCI interface. It delegates to app.snapshotManager if set.
-func (app *BaseApp) OfferSnapshot(req abci.RequestOfferSnapshot) abci.ResponseOfferSnapshot {
-	if app.snapshotManager == nil {
-		app.logger.Error("snapshot manager not configured")
-		return abci.ResponseOfferSnapshot{Result: abci.ResponseOfferSnapshot_ABORT}
-	}
+	var resp *abci.ResponseQuery
 
-	if req.Snapshot == nil {
-		app.logger.Error("received nil snapshot")
-		return abci.ResponseOfferSnapshot{Result: abci.ResponseOfferSnapshot_REJECT}
-	}
+	cmd, typ, arg := path[1], path[2], path[3]
+	switch cmd {
+	case "filter":
+		switch typ {
+		case "addr":
+			resp = app.FilterPeerByAddrPort(arg)
 
-	snapshot, err := snapshottypes.SnapshotFromABCI(req.Snapshot)
-	if err != nil {
-		app.logger.Error("failed to decode snapshot metadata", "err", err)
-		return abci.ResponseOfferSnapshot{Result: abci.ResponseOfferSnapshot_REJECT}
+		case "id":
+			resp = app.FilterPeerByID(arg)
+		}
+
+	default:
+		resp = sdkerrors.QueryResult(errorsmod.Wrap(sdkerrors.ErrUnknownRequest, "expected second parameter to be 'filter'"), app.trace)
 	}
 
-	err = app.snapshotManager.Restore(snapshot)
-	switch {
-	case err == nil:
-		return abci.ResponseOfferSnapshot{Result: abci.ResponseOfferSnapshot_ACCEPT}
+	return resp
+}
 
-	case errors.Is(err, snapshottypes.ErrUnknownFormat):
-		return abci.ResponseOfferSnapshot{Result: abci.ResponseOfferSnapshot_REJECT_FORMAT}
+// SplitABCIQueryPath splits a string path using the delimiter '/'.
+//
+// e.g. "this/is/funny" becomes []string{"this", "is", "funny"}
+func SplitABCIQueryPath(requestPath string) (path []string) {
+	path = strings.Split(requestPath, "/")
 
-	case errors.Is(err, snapshottypes.ErrInvalidMetadata):
-		app.logger.Error(
-			"rejecting invalid snapshot",
-			"height", req.Snapshot.Height,
-			"format", req.Snapshot.Format,
-			"err", err,
-		)
-		return abci.ResponseOfferSnapshot{Result: abci.ResponseOfferSnapshot_REJECT}
+	// first element is empty string
+	if len(path) > 0 && path[0] == "" {
+		path = path[1:]
+	}
 
-	default:
-		app.logger.Error(
-			"failed to restore snapshot",
-			"height", req.Snapshot.Height,
-			"format", req.Snapshot.Format,
-			"err", err,
-		)
+	return path
```

Comment:
And the implementation could be reduced to `strings.Split(strings.TrimPrefix(requestPath, "/"), "/")`.

Reply:
https://github.com/cosmos/cosmos-sdk/pull/15519#discussion_r1160228709